chrome.action is undefined if the action key in the manifest is missing or empty

elijahsawyers opened the pull request:
#### e880d73dbb4dc6c60f4b36486962a3ffba9e5c42
<pre>
chrome.action is undefined if the action key in the manifest is missing or empty
<a href="https://bugs.webkit.org/show_bug.cgi?id=270814">https://bugs.webkit.org/show_bug.cgi?id=270814</a>
<a href="https://rdar.apple.com/122330191">rdar://122330191</a>

Reviewed by Timothy Hatcher.

All keys in the dictionary in the manifest for action, browser_action, and
page_action should be optional. This patch makes the necessary changes for
that to be true.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::hasAction):
(WebKit::WebExtension::hasBrowserAction):
(WebKit::WebExtension::hasPageAction):
(WebKit::WebExtension::populateActionPropertiesIfNeeded):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed):

Canonical link: <a href="https://commits.webkit.org/275991@main">https://commits.webkit.org/275991@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/044b6f1f2df853f57e3d33dcee116f57d80ecf9d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43441 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22472 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45852 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46076 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39567 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45745 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26254 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19890 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35914 "Failure limit exceed. At least found 1 new test failure: fast/css-custom-paint/out-of-memory-while-adding-worklet-module.html (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44015 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19507 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37420 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16890 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17078 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38483 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1497 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/39627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38792 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47620 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18450 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15105 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42696 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19893 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41361 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20073 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5913 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19524 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon reviewed (changes requested):
You should add a test or two in `TEST(WKWebExtension, ActionParsing)`, `WKWebExtension.mm` for this case so it does not regress.

Also add a runtime test in `WKWebExtensionAPIAction.mm`, which would look something like this:

```objc
TEST(WKWebExtensionAPIAction, EmptyAction)
{
    static auto *actionManifest = @{
        @"manifest_version": @3,

        @"name": @"Test Action",
        @"description": @"Test Action",
        @"version": @"1.0",

        @"background": @{
            @"scripts": @[ @"background.js" ],
            @"type": @"module",
            @"persistent": @NO,
        },

        @"action": @{ }
    };

    auto *backgroundScript = Util::constructScript(@[
        @"if (browser.action) {",
        @"  browser.test.notifyPass()",
        @"} else {",
        @"  browser.test.notifyFail('browser.action should be defined when it is empty.')",
        @"}"
    ]);

    Util::loadAndRunExtension(actionManifest, @{ @"background.js": backgroundScript });
}
```

elijahsawyers commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
```diff
@@ -199,6 +199,12 @@
+
+    testManifestDictionary = @{ @"manifest_version": @3, @"name": @"Test", @"description": @"Test", @"version": @"1.0", @"action": @{ } };
+    testExtension = [[_WKWebExtension alloc] _initWithManifestDictionary:testManifestDictionary];
+    EXPECT_NS_EQUAL(testExtension.errors, @[ ]);
```
Are there any other assertions that we should add, or is having an empty action in the manifest and checking that there aren't any errors sufficient?

xeenon replied:
That looks good.

elijahsawyers committed:
chrome.action is undefined if the action key in the manifest is missing or empty
https://bugs.webkit.org/show_bug.cgi?id=270814
rdar://122330191

Reviewed by Timothy Hatcher.

All keys in the dictionary in the manifest for action, browser_action, and
page_action should be optional. This patch makes the necessary changes for
that to be true.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::hasAction):
(WebKit::WebExtension::hasBrowserAction):
(WebKit::WebExtension::hasPageAction):
(WebKit::WebExtension::populateActionPropertiesIfNeeded):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::isPropertyAllowed):

Canonical link: https://commits.webkit.org/275991@main
